### PR TITLE
[WIP] Adds types to Anchor

### PR DESCRIPTION
**Pull request by macalinao**

### CODE REVIEW

**Comment by macalinao** [comment]
cc @cheaney might be helping with this

**Comment by cheaney** [comment]
@armaniferrante I made the `escrow` example a typescript test. Let me now if you think we need more tests.

Also note that I needed to add a bash script (`create-idl-type.sh`) to generate the IDL type for the program. This was based on a script @macalinao shared with me. I think we'll need to add something similar to `anchor-cli` or `@project-serum/anchor` so that devs don't have to do this themselves.

File: examples/escrow/tests/escrow.ts [hunk 1]
```diff
@@ -1,20 +1,22 @@
...
+import {assert} from "chai";
+
+import {EscrowIDL} from "./types/escrow";
 
 describe("escrow", () => {
   const provider = anchor.Provider.env();
   anchor.setProvider(provider);
 
-  const program = anchor.workspace.Escrow;
+  const program = anchor.workspace.Escrow as anchor.Program<EscrowIDL>;
```
Comment (armaniferrante) [inline examples/escrow/tests/escrow.ts:11]: Hmm the account namespace doesn't type check. I can do `program.account.notAValidType` and it compiles.

Comment (cheaney) [reply examples/escrow/tests/escrow.ts:11]: Is this expected @macalinao or did I miss something.

Comment (macalinao) [reply examples/escrow/tests/escrow.ts:11]: I may have forgotten to add this in

Comment (cheaney) [reply examples/escrow/tests/escrow.ts:11]: I fixed this by [tweaking](https://github.com/project-serum/anchor/pull/537/files#diff-0e2c48e520eafaabcd0d9de371c7d2e74f5bbb4eba766216bb24efb265c0e5c3R68) the `AccountNamespace`. But for this to work, I also needed to customize the script that [produces](https://github.com/project-serum/anchor/pull/537/files#diff-9db4b8fb68cb3e81dd2ec7578b892eafc3ba7eb1db7535ecd5d41a76fee7fdbaR13) the idl type such that the Account names are camel case.

I think we'd want to think through the dev flow for producing the IDL type definition if we are to ship this. Maybe we add a script to `anchor-cli` that produces the IDL types for you?

Comment (cheaney) [reply examples/escrow/tests/escrow.ts:11]: Putting this on your radar @armaniferrante

**Comment by macalinao** [comment]
To anyone here: I've published a package here: https://github.com/saber-hq/saber-common/tree/master/packages/anchor-contrib

This contains most of these types. It's undocumented but feel free to ping me on Telegram (@imacalinao) if you're using it!

File: examples/escrow/tests/escrow.ts (hunk 1 above)
Comment (armaniferrante) [reply examples/escrow/tests/escrow.ts:11]: Is this ready for another review/testing?

**Comment by ChewingGlass** [comment]
@macalinao Great work on this PR! What needs to happen to kick it across the line? I'm ready to help out where needed

**Comment by armaniferrante** [comment]
> @macalinao Great work on this PR! What needs to happen to kick it across the line? I'm ready to help out where needed

@crispheaney is this WIP still or is it ready to go for another round of testing?

**Comment by crispheaney** [comment]
> > @macalinao Great work on this PR! What needs to happen to kick it across the line? I'm ready to help out where needed
> 
> @crispheaney is this WIP still or is it ready to go for another round of testing?

I fixed the `AccountNamespace` bug you surfaced, so in that sense it's ready to go. 

Before we merge this in, I think we'd want o define the user flow for taking their target's idl and transforming it the IDL type. I made a [hacky script](https://github.com/project-serum/anchor/pull/537/files#diff-9db4b8fb68cb3e81dd2ec7578b892eafc3ba7eb1db7535ecd5d41a76fee7fdbaR13) that creates the IDL type.

**Comment by ChewingGlass** [comment]
> Before we merge this in, I think we'd want to define the dev flow for taking their target's IDL and transforming it the IDL type. I made a [hacky script](https://github.com/project-serum/anchor/pull/537/files#diff-9db4b8fb68cb3e81dd2ec7578b892eafc3ba7eb1db7535ecd5d41a76fee7fdbaR13) that does this.


As someone consuming this, I think the hacky way is fine. Not everyone will want their types in the same place, I wrote our own custom script because we have a lerna setup and I want the generated code in its respective package. I think documenting it well would suffice for now, then automate later.

File: ts/src/program/namespace/types.ts [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+    (FieldsOfType<I> & { name: F })["type"],
+    Defined
+  >;
+};
+
+type TypeDefDictionary<T extends IdlTypeDef[], Defined> = {
+  [K in T[number]["name"]]: TypeDef<T[number] & { name: K }, Defined>;
+};
+
+export type IdlTypes<T extends Idl> = TypeDefDictionary<
```
Comment (ChewingGlass) [inline ts/src/program/namespace/types.ts:117]: Also useful to add and export:

```typescript
export type IdlAccounts<T extends Idl> = TypeDefDictionary<
  NonNullable<T["accounts"]>,
  Record<string, never>
>;
```

**Review by ChewingGlass — CHANGES_REQUESTED** [review]

**Comment by ChewingGlass** [comment]
This has a bunch of conflicts now, I'm going to create a separate PR since I did the merge. https://github.com/project-serum/anchor/pull/795

**Comment by Henry-E** [comment]
Yeh, this seems like a dead PR. Please use anchor typescript or anchor-client-gen for a static typescript client.

**Pull request closed without merging**